### PR TITLE
[routing] Do not show turn lanes without recommendations

### PR DESCRIPTION
**Pull request by nvarner**
This takes two major approaches towards resolving #4873.

- Fix the bug reported on Telegram/[Matrix](https://matrix.to/#/!wqHDOeaqKhfWGKYfGs:matrix.org/$BBSQ3UtuxcpEX2GXLv_Dzhk46xFnmaWGOxndfVc-sbw?via=matrix.org&via=t2bot.io&via=tchncs.de) that occurred when lanes were tagged as having no turn restrictions, but not recommended to users
  <details>
  <summary>see screenshot from bug report and after fix</summary>
  ### Bug when reported

  screenshot of navigation with two greyed out lane suggestions
  
  ### Now
  Screenshot_20230330_011029
  </details>

- With complex intersections where we still cannot find a lane to recommend, do not show the lane guidance box
  <details>
    <summary>see screenshot from after fix</summary>

    Screenshot_20230330_002637
  </details>

Notably, this PR does not implement some mentioned features:
- Distance or countdown to turn. The turn is the same as the one that already has the distance countdown on the left, so another counter would be redundant.
- Arrows that show one lane with multiple turn directions. I can implement this, but it would need new icons. Creating them would take some time, and I would like to fix this issue as quickly as possible.
- A unified, cleaner top navigation UI. I have a prototype for this, but it's not yet ready. Again, completing it will take some time, and this can be resolved without it.

### CODE REVIEW

**Review by vng — COMMENTED** [review]
I will take a deeper look later, but before that - remove cameras_to_ways.bin and kothic from this branch!

File: routing/turns.cpp [hunk 1]
```diff
@@ -221,6 +221,14 @@ bool IsLeftOrRightTurn(CarDirection t)
   return IsLeftTurn(t) || IsRightTurn(t);
 }
 
+bool IsTurnMadeFromLeft(CarDirection t) {
```
Comment (vng) [inline routing/turns.cpp:224]: nit: { braces from the new line

File: routing/turns.cpp [hunk 2]
```diff
@@ -221,6 +221,14 @@ bool IsLeftOrRightTurn(CarDirection t)
...
+  return IsLeftTurn(t) || t == CarDirection::UTurnLeft || t == CarDirection::ExitHighwayToLeft;
+}
+
+bool IsTurnMadeFromRight(CarDirection t) {
```
Comment (vng) [inline routing/turns.cpp:228]: nit: nit: { braces from the new line

File: routing/turns.cpp [hunk 3]
```diff
@@ -290,6 +298,11 @@ bool IsLaneWayConformedTurnDirectionApproximately(LaneWay l, CarDirection t)
   }
 }
 
+bool IsLaneUnrestricted(const SingleLaneInfo & lane)
+{
+  return lane.m_lane.size() == 1 && lane.m_lane[0] == LaneWay::None;
```
Comment (vng) [inline routing/turns.cpp:303]: For future: Is there any reason to store None single lane?

File: routing/car_directions.cpp [hunk 4]
```diff
@@ -622,21 +622,50 @@ bool FixupLaneSet(CarDirection turn, vector<SingleLaneInfo> & lanes,
   return isLaneConformed;
 }
 
+template <typename It>
+bool SelectFirstUnrestrictedLane(LaneWay direction, It lanesBegin, It lanesEnd)
+{
+  const It firstUnrestricted = find_if(lanesBegin, lanesEnd, IsLaneUnrestricted);
```
Comment (biodranik) [inline routing/car_directions.cpp:628]:
```suggestion
  const It firstUnrestricted = std::find_if(lanesBegin, lanesEnd, IsLaneUnrestricted);
```

File: routing/car_directions.cpp [hunk 5]
```diff
@@ -622,21 +622,50 @@ bool FixupLaneSet(CarDirection turn, vector<SingleLaneInfo> & lanes,
...
+  if (firstUnrestricted == lanesEnd)
+    return false;
+  
+  firstUnrestricted->m_isRecommended = true;
+  firstUnrestricted->m_lane.insert(firstUnrestricted->m_lane.begin(), direction);
+  return true;
+}
+
+bool SelectUnrestrictedLane(CarDirection turn, vector<SingleLaneInfo> & lanes)
```
Comment (biodranik) [inline routing/car_directions.cpp:637]:
```suggestion
bool SelectUnrestrictedLane(CarDirection turn, std::vector<SingleLaneInfo> & lanes)
```

File: routing/car_directions.cpp [hunk 6]
```diff
@@ -622,21 +622,50 @@ bool FixupLaneSet(CarDirection turn, vector<SingleLaneInfo> & lanes,
...
+{
+  if (IsTurnMadeFromLeft(turn))
+    return SelectFirstUnrestrictedLane(LaneWay::Left, lanes.begin(), lanes.end());
+  else if (IsTurnMadeFromRight(turn))
+    return SelectFirstUnrestrictedLane(LaneWay::Right, lanes.rbegin(), lanes.rend());
+  return false;
```
Comment (biodranik) [inline routing/car_directions.cpp:643]: Is some todo/descriptive comment needed here?

File: tools/kothic
Comment (biodranik) [inline tools/kothic:1]: Why is this change here?

**Comment by vng** [comment]
@nvarner Can you please fix the comments, or I can update your branch if you don't mind?

**Comment by nvarner** [comment]
I'm sorry I wasn't able to get to this sooner, but thank you for fixing it! I'm not sure how I managed to mess up my commits like that....
